bookkeeper-common/src/main/java/org/apache/bookkeeper/common/component/ComponentStarter.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.bookkeeper.common.component;
+
+import java.util.concurrent.CountDownLatch;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Utils to start components.
+ */
+@Slf4j
+public class ComponentStarter {
+
+    static class ComponentShutdownHook implements Runnable {
+
+        private final LifecycleComponent component;
+        private final CountDownLatch aliveLatch;
+
+        ComponentShutdownHook(LifecycleComponent component,
+                              CountDownLatch aliveLatch) {
+            this.component = component;
+            this.aliveLatch = aliveLatch;
+        }
+
+        @Override
+        public void run() {
+            aliveLatch.countDown();
+            log.info("Closing component {} in shutdown hook.", component.getName());
+            try {
+                component.close();
+                log.info("Closed component {} in shutdown hook successfully. Exiting.", component.getName());
+            } catch (Exception e) {
+                log.error("Failed to close component {} in shutdown hook gracefully, Exiting anyway",
+                    component.getName(), e);
+            }
+        }
+
+    }
+
+    /**
+     * Start a component and register a shutdown hook.
+     *
+     * @param component component to start.
+     */
+    public static void startComponent(LifecycleComponent component,
+                                      CountDownLatch aliveLatch) {
+        Runtime.getRuntime().addShutdownHook(new Thread(
```

Comment:
this is not good.
On unit tests for instance we are going to pollute the Java Runtime with shutdownhooks.
Shutdown hooks introduce a lot of troubles and in my experience I saw a lot of deadlocks.
I would like to drop this if there is no strong evidence of benefit.

Reply:
Got it. Do you think we can drop it anyway?

Reply:
Got it. Do you think we can drop it anyway?

Reply:
@eolivelli I am wondering why do you want to drop this, especially this is not a new hook introduced by this code. This shutdown hook is an attempt to do a clean shutdown, when the process was killed.

Reply:
Before this change the hook applies only to the bookie server. Now we are adding an hook for every component.
There will be no control on when these hooks will run and in which order.
We could run into deadlocks if there is no coordination of components and for their own pluggable nature we won't have this ability of coordination.

Maybe this is ok for the BookieServer in order to try to do a clean shutdown of journals and storage. This usually can bring better recovery.

The hook is not guaranteed to be executed,  so functionally we can drop it.

For tests we should disable it anyway because it induces a strong reference to the component and this is a memory leak. We don't suffer about this now because we are always forking new processes.

Sorry for typos, using mobile

Reply:
@eolivelli - 

1) this is just a util function to start a lifecycle component. You can use it or not use it. It is up to the caller.
2) the actual place calling this is a stack of lifecycle components, which is still exactly the bookie server. so there is no behavior changes
3) this function isn't actually called by other places except the main method starting the bookie server. so it doesn't run by any tests.

in this patch, we don't change any behavior of existing bookie server, how it starts and stops. we don't introduce new behavior and we don't run this for any components in tests. so I am not sure if we really want to drop it here.

Reply:
@sijie thanks
My bad I did not see this is just an utility method and how it was used.
I apologize I did not have my laptop to have a global view.

For me is ok